src/NLog/Layouts/Typed/Layout.cs

```diff
@@ -0,0 +1,350 @@
+﻿// 
+// Copyright (c) 2004-2020 Jaroslaw Kowalski <jaak@jkowalski.net>, Kim Christensen, Julian Verdurmen
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without 
+// modification, are permitted provided that the following conditions 
+// are met:
+// 
+// * Redistributions of source code must retain the above copyright notice, 
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Globalization;
+using System.Text;
+using JetBrains.Annotations;
+using NLog.Common;
+using NLog.Config;
+using NLog.Internal;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout with a simple value (e.g. int) or a layout which results in a simple value (e.g. ${counter})
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class Layout<T> : Layout, IRenderable<T>
+    {
+        // ReSharper disable StaticMemberInGenericType - this is safe, static ctor is called for every generic type
+        [NotNull] private static readonly Type Type;
+        [NotNull] private static readonly string TypeNamed;
+        // ReSharper restore StaticMemberInGenericType
+
+        [CanBeNull] private readonly Layout _layout;
```

Comment:
Because this nested `Layout` cannot be discovered, then `Layout<T>` must announce (and check) whether it is ThreadAgnostic or MutableUnsafe.

Reply:
Also need to override `PrecalculateBuilder` and call `_layout.PrecalculateBuilder` when necessary.

Reply:
Maybe we should add the same attributes as SimpleLayout? 
```c#
    [ThreadAgnostic]
    [ThreadSafe]
    [AppDomainFixedOutput]
    public class SimpleLayout : Layout, IUsesStackTrace
```

Reply:
> Also need to override `PrecalculateBuilder` and call `_layout.PrecalculateBuilder` when necessary.

Could you please do a code proposal for this?

Reply:
Something like this:

```c#
       protected override void InitializeLayout()
       {
            base.InitializeLayout();
            ThreadAgnostic = _layout?.ThreadAgnostic ?? true;
            MutableUnsafe = _layout?.MutableUnsafe ?? false;
       }

       internal override void PrecalculateBuilder(LogEventInfo logEvent, StringBuilder target)
       {
           _layout?.PrecalculateBuilder(logEvent, target);
       }
```

Reply:
done